lib/srv/db/redis/protocol.go

```diff
@@ -0,0 +1,257 @@
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+
+package redis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common/role"
+	"github.com/gravitational/trace"
+)
+
+// processCmd processes commands received from connected client. Most commands are just passed to Redis instance,
+// but some require special actions:
+//  * Redis 7.0+ commands are rejected as at the moment of writing Redis 7.0 hasn't been released and go-redis doesn't support it.
+//  * RESP3 commands are rejected as Teleport currently doesn't support this version of protocol.
+//  * Subscribe related commands created a new DB connection as they change Redis request-response model to Pub/Sub.
+func (e *Engine) processCmd(ctx context.Context, redisClient redis.UniversalClient, cmd *redis.Cmd) error {
+	switch strings.ToLower(cmd.Name()) {
+	case "hello":
+		return redis.RedisError("RESP3 is not supported")
+	case "auth":
+		return e.processAuth(ctx, redisClient, cmd)
+	case "subscribe":
+		return e.subscribeCmd(ctx, redisClient.Subscribe, cmd)
+	case "psubscribe":
+		return e.subscribeCmd(ctx, redisClient.PSubscribe, cmd)
+	case "punsubscribe":
+		// TODO(jakub): go-redis doesn't expose any API for this command. Investigate alternative options.
+		return errors.New("PUNSUBSCRIBE is not supported by Teleport")
+	case "ssubscribe", "sunsubscribe":
+		// go-redis doesn't support Redis 7+ commands yet.
+		return errors.New("Redis 7.0+ is not yet supported")
+	default:
+		// Here the command is sent to the DB.
+		return redisClient.Process(ctx, cmd)
+	}
+}
+
+type redisSubscribeFn = func(ctx context.Context, channels ...string) *redis.PubSub
+
+// subscribeCmd handles subscription in Redis. A new connection is created and maintained to read published messages
+// from Redis and send them back to the client.
+func (e *Engine) subscribeCmd(ctx context.Context, subscribeFn redisSubscribeFn, cmd *redis.Cmd) error {
+	if len(cmd.Args()) < 2 {
+		return redis.RedisError("invalid command")
+	}
+
+	args := make([]string, 0, len(cmd.Args()))
+	for _, a := range cmd.Args()[1:] {
+		args = append(args, a.(string))
+	}
+	pubSub := subscribeFn(ctx, args...)
+
+	return e.processPubSub(ctx, pubSub)
+}
+
+// processPubSub reads messages from Redis and forward them to connected client.
+func (e *Engine) processPubSub(ctx context.Context, pubSub *redis.PubSub) error {
```

Comment:
if `pubSub := subscribeFn(ctx, args...)`  creates a new client we should make sure that it will close when a on main client disconnect and properly release resources.

Reply:
Good catch. I'm pretty sure that go-redis closes that connection anyways when the main connection is being closed, but I agree what we should call close here too. Added